### PR TITLE
GUI for GSettings

### PR DESCRIPTION
**Pull request by sei0o**
I'm working on #142 and trying to create the UI first (see `components/settings/`) then to integrate it with some structs in `src/settings.rs`.

Currently I am stuck at how to handle the `AdwPreferencesWindow`. I have set up XML file with some widgets, added a button in the user menu, but an empty window is shown: 

image

### CODE REVIEW

**Comment by xou816** [comment]
Hi! Nice, thank you for working on this!

Haven't played with the Preferences stuff yet, but looks like you might need to have a PreferencesPage as a child of that PreferencesWindow first

(...also, the `<child>` tag should only have one `<object>` child!)

**Comment by jannuary** [comment]
You need to have the settings.ui file defined in the gresource schema, like this:
```
<?xml version="1.0" encoding="UTF-8"?>
<gresources>
  <gresource prefix="/dev/alextren/Spot">
...
    <file alias="components/settings.ui">app/components/settings/settings.ui</file>
...
  </gresource>
</gresources>
```

**Comment by sei0o** [comment]
It is working after applying your advice and some other tweaks. Thanks!
image

**Comment by sei0o** [comment]
So now the settings window is working and the updated GSettings properties are loaded on next startup. 

I would like to make they are reloaded during the execution of Spot, as many applications do. Based on my understanding, players and some other components make use of `SpotSettings`. On startup,  the `SpotSettings` is `clone`d & passed down on those components. Therefore they are not synchronized with the original one or GSettings. 

My idea is to put a `SpotSettings` in the `AppState` and configure necessary signals/actions to make reloading happen while the app is running (the design doc is quite helpful!). What do you think?

**Comment by xou816** [comment]
Neat!

Given that most if not all settings require a restart, it is not that big of a deal if we don't manage these settings through the shared AppState -- but you can absolutely do that if you want :)

As for persisting those settings, there's an example of that in settings.rs -- you could do it from pretty much anywhere I think.

**Comment by sei0o** [comment]
I've just pushed the latest commit, which allows us to apply new settings without restarting. The problem, however, is that the window cannot be closed twice. This bug has been there since I created an empty PreferencesWindow and is pretty weird...

**Comment by jannuary** [comment]
> I've just pushed the latest commit, which allows us to apply new settings without restarting. The problem, however, is that the window cannot be closed twice. This bug has been there since I created an empty PreferencesWindow and is pretty weird... 

It's a known issue, you need to have hide-on-close enabled on the PreferencesWindow.

File: src/app/components/user_menu/user_menu.rs
Comment (jannuary) [inline src/app/components/user_menu/user_menu.rs:65]: ~~Settings~~ Preferences

File: src/app/components/settings/settings.ui
Comment (jannuary) [inline src/app/components/settings/settings.ui:28]: This should be hidden if ALSA isn't used.

File: src/app/components/settings/settings.ui
Comment (jannuary) [inline src/app/components/settings/settings.ui:37]: I think it makes more sense to expose this setting as something like "Audio Quality" with Low/Medium/High options, just like Spotify itself does.

File: src/app/components/settings/settings.ui
Comment (jannuary) [inline src/app/components/settings/settings.ui:7]: We don't have nearly enough options to justify it, so disable the search:
```suggestion
    <property name="hide-on-close">1</property>
    <property name="search-enabled">0</property>
```

File: src/app/components/settings/settings.ui
Comment (jannuary) [inline src/app/components/settings/settings.ui:31]:
```suggestion
                  <object class="GtkEntry" id="alsa_device">
                      <property name="valign">center</property>
                  </object>
```
Otherwise it looks weirdly stretched.

File: src/app/components/settings/settings.ui
Comment (jannuary) [inline src/app/components/settings/settings.ui:77]: Same `valign` as the other entry.

**Review by jannuary — CHANGES_REQUESTED** [review]

**Comment by xou816** [comment]
Hi! been a little busy but I'll have a look :)

File: src/app/state/settings_state.rs
Comment (xou816) [inline src/app/state/settings_state.rs:27]: I think  you could store the Settings here as mentioned at some point :)

That way, the action could have no payload with it (just `ChangePlayerSettings`) and whenever that action is dispatched, we just do `self.settings = SpotSettings::new_from_gsettings()` since it's already synced probably? 

that way you don't have to do the mapping from the Adwaita widgets to SpotSettings everytime

File: src/app/components/settings/settings.rs
Comment (xou816) [inline src/app/components/settings/settings.rs:259]: think you can upcast_ref to a window in one go, no need to cast to a preferences window first

File: src/app/components/settings/settings.rs
Comment (xou816) [inline src/app/components/settings/settings.rs:182]: think you can upcast_ref to a window in one go, no need to cast to a preferences window first

File: src/player/player.rs
Comment (xou816) [inline src/player/player.rs:54]: I think (I need to double check) that the derived PartialEq impl won't check for equality on the nested String here, so there would be Alsa("foo") == Alsa("bar")

**Review by xou816 — COMMENTED** [review]
nice work!! I'll try to be more reactive to questions and feedback as you go through the review!

File: src/app/components/settings/settings.ui
Comment (sei0o) [reply src/app/components/settings/settings.ui:37]: Agreed. Current bitrate options (96/160/320 kbps) correspond to [Normal/High/"Very high" in Spotify](https://support.spotify.com/us/article/audio-quality/), so it would be better to reflect that in our preferences?

File: src/player/player.rs
Comment (sei0o) [reply src/player/player.rs:54]: Double checked on [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2eb03c1532fce3300c3aa1e7a32f6d01). `derive` handles it well I suppose :)

File: src/app/state/settings_state.rs
Comment (sei0o) [reply src/app/state/settings_state.rs]: Makes sense. Moved SpotSettings into SettingsState and SettingsWindow use it whenever possible, while there are still some redundant `SpotSettings::new_from_gsettings()` calls from the player and `update_with` handlers for SettingsState.

Comment (xou816) [reply src/app/state/settings_state.rs]: I am nitpicking a bit, but I think one action is enough here, and the state should decide whether to return no event or `vec![SettingsEvent::PlayerSettingsChanged]` if the player part of the settings have changed

Comment (sei0o) [reply src/app/state/settings_state.rs]: `set_player_settings` still has to emit `PlaybackAction::Stop` to stop playing (otherwise there will be some inconsistency in UI), so the two same `if`s are put in `update_with` and `Settings::new`. It's not a nitpick because I had the same concern, just thought of avoiding the duplicate ifs at first.

File: src/app/state/settings_state.rs
Comment (xou816) [inline src/app/state/settings_state.rs:40]:
```diff
     fn update_with(&mut self, action: std::borrow::Cow<Self::Action>) -> Vec<Self::Event> {
         match action.into_owned() {
             SettingsAction::ChangeSettings => {
-                let old_settings = self.settings.clone();
-                self.settings = SpotSettings::new_from_gsettings().unwrap_or_default();
-                if self.settings.player_settings == old_settings.player_settings {
-                    vec![]
-                } else {
+                let old_settings = &self.settings;
+                let new_settings = SpotSettings::new_from_gsettings().unwrap_or_default();
+                let player_settings_changed =
+                    new_settings.player_settings != old_settings.player_settings;
+                self.settings = new_settings;
+                if player_settings_changed {
                     vec![SettingsEvent::PlayerSettingsChanged.into()]
+                } else {
+                    vec![]
                 }
             }
         }
```

(you can avoid the extra clone here)

File: src/app/state/settings_state.rs
Comment (xou816) [reply src/app/state/settings_state.rs]: Oh you're right absolutely!

**Comment by Diegovsky** [comment]
Great work! I can't wait for this to me merged :)

File: src/app/state/settings_state.rs
Comment (sei0o) [reply src/app/state/settings_state.rs]: Fixed in the latest commit :)

**Comment by xou816** [comment]
Thank you for your work and patience through the review! Merging this :)

**Comment by sei0o** [comment]
Learned a lot from this PR... I really appreciate reviews & advice!

**Pull request merged**
